lighthouse-core/audits/seo/canonical.js

```diff
@@ -0,0 +1,167 @@
+  }
+}
+
+/**
+ * Returns a primary domain for provided URL (e.g. http://www.example.com -> example.com).
+ * Note that it does not take second-level domains into account (.co.uk).
+ * @param {URL} url
+ * @returns {string}
+ */
+function getPrimaryDomain(url) {
+  return url.hostname.split('.').slice(-2).join('.');
+}
+
+class Canonical extends Audit {
+  /**
+   * @return {!AuditMeta}
+   */
+  static get meta() {
+    return {
+      name: 'canonical',
+      description: 'Document has a valid `rel=canonical`',
+      failureDescription: 'Document does not have a valid `rel=canonical`',
+      helpText: 'Canonical links suggest which URL to show in search results. ' +
+        'Read more in [Use canonical URLs]' +
+        '(https://support.google.com/webmasters/answer/139066).',
+      requiredArtifacts: ['Canonical', 'Hreflang'],
+    };
+  }
+
+  /**
+   * @param {!Artifacts} artifacts
+   * @return {!AuditResult}
+   */
+  static audit(artifacts) {
+    const devtoolsLogs = artifacts.devtoolsLogs[Audit.DEFAULT_PASS];
+
+    return artifacts.requestMainResource(devtoolsLogs)
+      .then(mainResource => {
+        const baseURL = new URL(mainResource.url);
+        let canonicals = [];
+        let hreflangs = [];
+
+        mainResource.responseHeaders
+          .filter(h => h.name.toLowerCase() === LINK_HEADER)
+          .forEach(h => {
+            canonicals = canonicals.concat(getCanonicalLinksFromHeader(h.value));
+            hreflangs = hreflangs.concat(getHreflangsFromHeader(h.value));
+          });
+
+        canonicals = canonicals.concat(artifacts.Canonical);
+
+        artifacts.Hreflang.forEach(({href}) => hreflangs.push(href));
+
+        hreflangs = hreflangs
+          .filter(href => isValidRelativeOrAbsoluteURL(href))
+          .map(href => (new URL(href, baseURL)).href); // normalize URLs
+
+        if (canonicals.length === 0) {
+          return {
+            rawValue: true,
+            notApplicable: true,
+          };
+        }
+
+        if (canonicals.length > 1) {
+          return {
+            rawValue: false,
+            displayValue: `multiple URLs (${canonicals.join(', ')})`,
+          };
+        }
+
+        const canonical = canonicals[0];
+
+        if (!isValidRelativeOrAbsoluteURL(canonical)) {
+          return {
+            rawValue: false,
+            displayValue: `invalid URL (${canonical})`,
+          };
+        }
+
+        if (!URL.isValid(canonical)) {
+          return {
+            rawValue: false,
+            displayValue: `relative URL (${canonical})`,
+          };
+        }
+
+        const canonicalURL = new URL(canonical);
+
+        // cross-language or cross-country canonicals are a common issue
+        if (hreflangs.includes(baseURL.href) && hreflangs.includes(canonicalURL.href) &&
+          baseURL.href !== canonicalURL.href) {
+          return {
+            rawValue: false,
+            displayValue: `points to another hreflang location (${baseURL.href})`,
```

Comment:
Rather than overloading displayValue I think we should abuse debugString like we already do for the PWA ones.

We can even extend https://github.com/GoogleChrome/lighthouse/blob/master/lighthouse-core/audits/multi-check-audit.js much like the splash screen one.

(yes, debugString is supposed to be exceptions, not details of failing tests. But since the PWA audits are already violating that, we might as well violate it here, too.)

Reply:
TIL about the multi-check-audit! Looks cool, but it probably makes most sense for audits returning multiple failures (`Failures: ${result.failures.join(', ')}.`) and this one doesn't, so I'll just switch `displayValue`s to `debugString`s.